zerver/lib/export.py

```diff
@@ -0,0 +1,653 @@
+    user_recipient_ids = set(x["id"] for x in user_recipients)
+    user_subscription_query = Subscription.objects.filter(user_profile__in=user_profile_ids,
+                                                          recipient_id__in=user_recipient_ids)
+    user_subscription_dicts = [model_to_dict(x) for x in user_subscription_query]
+
+    response["zerver_userpresence"] = [model_to_dict(x) for x in
+                                       UserPresence.objects.filter(user_profile__in=user_profile_ids)]
+    floatify_datetime_fields(response, 'zerver_userpresence', 'timestamp')
+
+    response["zerver_useractivity"] = [model_to_dict(x) for x in
+                                       UserActivity.objects.filter(user_profile__in=user_profile_ids)]
+    floatify_datetime_fields(response, 'zerver_useractivity', 'last_visit')
+
+    response["zerver_useractivityinterval"] = [model_to_dict(x) for x in
+                                               UserActivityInterval.objects.filter(user_profile__in=user_profile_ids)]
+    floatify_datetime_fields(response, 'zerver_useractivityinterval', 'start')
+    floatify_datetime_fields(response, 'zerver_useractivityinterval', 'end')
+
+    stream_query = Stream.objects.filter(realm=realm)
+    if not include_invite_only:
+        stream_query = stream_query.filter(invite_only=False)
+    response['zerver_stream'] = [model_to_dict(x, exclude=["email_token"]) for x in stream_query]
+    floatify_datetime_fields(response, 'zerver_stream', 'date_created')
+    stream_ids = set(x["id"] for x in response['zerver_stream'])
+    stream_recipients = [model_to_dict(x)
+                         for x in Recipient.objects.filter(type=2, type_id__in=stream_ids)]
+
+    stream_recipient_ids = set(x["id"] for x in stream_recipients)
+
+    stream_subscription_query = Subscription.objects.filter(user_profile__in=user_profile_ids,
+                                                            recipient_id__in=stream_recipient_ids)
+    stream_subscription_dicts = [model_to_dict(x) for x in stream_subscription_query]
+
+    if include_private:
+        # First we get all huddles involving someone in the realm.
+        realm_huddle_subs = Subscription.objects.select_related("recipient").filter(recipient__type=3,
+                                                                                    user_profile__in=user_profile_ids)
+        realm_huddle_recipient_ids = set(sub.recipient_id for sub in realm_huddle_subs)
+
+        # Mark all Huddles whose recipient ID contains a cross-realm user.
+        unsafe_huddle_recipient_ids = set()
+        for sub in Subscription.objects.select_related().filter(recipient__in=realm_huddle_recipient_ids):
+            if sub.user_profile.realm != realm:
+                # In almost every case the other realm will be zulip.com
+                unsafe_huddle_recipient_ids.add(sub.recipient_id)
+
+        # Now filter down to just those huddles that are entirely within the realm.
```

Comment:
Cool, added this:
        # This is important for ensuring that the User objects needed                                                   
        # to import it on the other end exist (since we're only                                                         
        # exporting the users from this realm), at the cost of losing                                                   
        # some of these cross-realm messages.